flatironinstitute/nomad
# Add infrastructure for per-iteration data output.

Adds a hook for doing per-iteration data file output, and functions to provide a reasonably-safe directory for such output & execute the resulting callback at every iteration of the main decomposition loop.

Closes #7.

### Type of change
- [ ] Bug fix (non-breaking change which fixes an unintended
or erroneous behavior)
- [x] New features (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that changes how existing
functionality works, other than correcting errors)
- [ ] Code improvement (no new functionality, but improving project structure)
- [ ] Documentation only (changes that improve or expand instructions
to users, without changing program behavior)


## Motivation and Context
Some kernels (involving e.g. momentum) may require per-iteration diagnostic output to successfully implement and debug. The challenge is integrating this output into the main decomposition loop in a way that:
- Minimizes implementation effort for kernel developers;
- Provides consistent control (across kernels) of the resulting data logs; and
- Does not require any extra implementation for kernels that don't take advantage of the functionality.

This PR accomplishes that aim by adding a (non-abstract) method to the kernel base class, defining configuration options for per-iteration diagnostic output, and providing centralized functionality for directory management.

## Description

This PR introduces the following changes:

* Add `per_iteration_diagnostic()` method to base kernel class `KernelBase`
  * The method takes a requested logging level (defined through `enum`) and a directory where output can take place; unless overridden, the directory is guaranteed to be timestamped with the runtime and empty at the time of run start
  * The method has a base implementation; for child classes that do not override, the default is to log a warning if the function is actually called
  * Kernel classes that actually have a functional implementation of this method will also have access to all other information attached to the `Kernel` object, including iteration count.
  * Affected files: [src/fi_nomad/kernels/kernel_base.py](https://github.com/flatironinstitute/nomad/blob/7-per-iteration-diagnostic-hook/src/fi_nomad/kernels/kernel_base.py), [test/test_kernels/test_base_model_free_kernel.py](https://github.com/flatironinstitute/nomad/blob/7-per-iteration-diagnostic-hook/test/test_kernels/test_base_model_free_kernel.py) (for testing)
* Implement enumerations and types to describe per-iteration diagnostics
  * Added enum for levels of diagnostic detail, with included space for future refinement
  * Added data class for per-loop diagnostic output configuration info
  * [src/fin_nomad/types/\_\_init\_\_.py](https://github.com/flatironinstitute/nomad/blob/7-per-iteration-diagnostic-hook/src/fi_nomad/types/__init__.py), [src/fi_nomad/types/enums.py](https://github.com/flatironinstitute/nomad/blob/7-per-iteration-diagnostic-hook/src/fi_nomad/types/enums.py), [src/fi_nomad/types/types.py](https://github.com/flatironinstitute/nomad/blob/7-per-iteration-diagnostic-hook/src/fi_nomad/types/types.py) 
* Added utilities for safely managing diagnostic-information output directories.
  * Utility provides a new-created, empty, timestamped directory beneath a user-requested base directory. This can be used for all output per run.
  * If a corresponding timestamped directory already exists for some reason, we add a `-x` counter until we find an empty directory. (This is expected to be unlikely.)
  * User has the option to turn off the timestamping and subdirectory feature and just use exactly the base directory as requested, but the utility will still throw an exception if the input base path is a file.
  * Affected files: [src/fi_nomad/util/path_util.py](https://github.com/flatironinstitute/nomad/blob/7-per-iteration-diagnostic-hook/src/fi_nomad/util/path_util.py), [test/test_util/test_path_util.py](https://github.com/flatironinstitute/nomad/blob/7-per-iteration-diagnostic-hook/test/test_util/test_path_util.py) (testing)
* Main loop generates a callback for the configured per-iteration output for the kernel.
  * Factory utility handles generating this callback from a configured kernel
  * Note that this takes the form of a callback that's [closed](https://en.wikipedia.org/wiki/Closure_(computer_programming)) over the configuration settings--we will create an output directory once, and the entire run will use the diagnostic level requested at the beginning of execution.
  * Affected files: [src/fi_nomad/entry.py](), [src/fi_nomad/util/factory_util.py](https://github.com/flatironinstitute/nomad/blob/7-per-iteration-diagnostic-hook/src/fi_nomad/util/factory_util.py), [test/test_decompose_loop.py]() (testing that callback is called), [test/test_util/test_factory_util.py](https://github.com/flatironinstitute/nomad/blob/7-per-iteration-diagnostic-hook/test/test_util/test_factory_util.py) (testing behavior of callback generation), 
* Incidental change: offloaded the post-run summarization and timing report to its own function within `entry.py` to keep the main loop cleaner
* Incidental change: Converted `Union[x, None]` typing throughout the library to `Optional[x]`
* Updates to README explaining how to call the new functionality (once implemented)
* Update to `pyproject.toml` to silence a spurious `pylint` error

## Testing

The PR adds passing unit tests for the following properties:
* Kernel base implements a warn-only per-iteration diagnostic (tested via simplest implementing class: `test/test_kernels/test_base_model_free_kernel.test_base_model_free_kernel_per_iteration_diagnostic_does_nothing`, inspect log contents)
* Test output directory configuration (`test/test_util/test_path_util.py`):
  * Ensure timestamps are created as expected
  * Ensure error is thrown if requested base path is an existing file
  * Ensure function returns requested base if `use_exact_path` is set
  * Ensure function creates a requested basepath if it does not exist, and does not create one that does exist
  * Test incrementing the counter if the intended output directory already exists
* Test callback configuration (`test/test_util/test_factory_util.py`):
  * Ensure function returns a no-op callback if no configuration data was passed
  * Ensure function calls the kernel's defined code if valid diagnostic configuration was passed
* main `decompose` loop calls diagnostic callback (`test/test_decompose_loop.test_decompose_calls_diagnostic_callback`): add mock for the callback and verify that it is called the same number of times as the manual iterations setting.
* We now check that the offloaded `entry.do_final_report()` produces timing strings as expected (examine log output in `test/test_decompose_loop.test_do_final_report_computes_time()`)

Since there is no observable behavior change for any existing kernels, no changes are made to the existing integration tests.


## Checklist
  - [x] I've run the automatic formatters and made any needed changes
  - [x] I've ensured that my new code is explained in docstrings/comments
- [x] My submission requires a change to the documentation
  - [x] I've updated the documentation to reflect the changes
- [x] I've added unit tests that cover the visible behavior of the changes
- [ ] (**N/A**) I've added or modified integration tests that ensure the whole system
works on a practical problem after my changes
